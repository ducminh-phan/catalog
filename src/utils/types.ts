export type PaginationParams = {
  offset: number;
  limit: number;
};

export type Category = {
  id: number;
  name: string;
  description: string;
};
