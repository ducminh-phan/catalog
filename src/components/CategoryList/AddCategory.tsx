import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from "@material-ui/core";
import AddIcon from "@material-ui/icons/Add";
import { unwrapResult } from "@reduxjs/toolkit";
import { makeRequired, makeValidate, TextField } from "mui-rff";
import React, { ReactNode } from "react";
import { Form } from "react-final-form";
import { connect, ConnectedProps } from "react-redux";
import * as yup from "yup";

import { addCategory } from "actions/category";
import { useNotification } from "contexts/notification";

const schema = yup.object().shape({
  name: yup.string().required().min(5),
  description: yup.string().required(),
});

const validate = makeValidate(schema);

const required = makeRequired(schema);

type FormData = yup.InferType<typeof schema>;

const mapDispatchToProps = {
  addCategory,
};

const connector = connect(null, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

type Props = PropsFromRedux;

const AddCategory = (props: Props): React.ReactElement => {
  // eslint-disable-next-line no-shadow
  const { addCategory } = props;
  const [open, setOpen] = React.useState(false);
  const { setNotification } = useNotification();

  const handleOpen = (): void => {
    setOpen(true);
  };

  const handleClose = (): void => {
    setOpen(false);
  };

  const handleAdd = async (data: FormData): Promise<void> => {
    const resultAction = await addCategory(data);

    handleClose();

    try {
      unwrapResult(resultAction);
      setNotification({
        severity: "success",
        message: "Add category successfully",
      });
    } catch (e) {
      setNotification({
        severity: "error",
        message: e.message ?? "Add category failed",
      });
    }
  };

  return (
    <>
      <Button color="inherit" onClick={handleOpen}>
        <AddIcon />
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-labelledby="category-add-form-title"
        fullWidth
      >
        <DialogTitle id="category-add-form-title">Add Category</DialogTitle>
        <DialogContent>
          <Form
            onSubmit={handleAdd}
            validate={validate}
            render={({ handleSubmit, invalid }): ReactNode => (
              <form onSubmit={handleSubmit} noValidate>
                <TextField label="Name" name="name" required={required.name} />
                <TextField
                  label="Description"
                  name="description"
                  required={required.description}
                />
                <DialogActions>
                  <Button type="button" onClick={handleClose}>
                    Cancel
                  </Button>
                  <Button color="primary" type="submit" disabled={invalid}>
                    Submit
                  </Button>
                </DialogActions>
              </form>
            )}
          />
        </DialogContent>
      </Dialog>
    </>
  );
};

export default connector(AddCategory);
