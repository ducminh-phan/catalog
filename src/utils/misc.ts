export type JsonPrimitive = string | number | boolean | null;
export type JsonMap = {
  [member: string]: JsonPrimitive | JsonArray | JsonMap;
};
export type JsonArray = Array<JsonPrimitive | JsonArray | JsonMap>;
export type Json = JsonPrimitive | JsonMap | JsonArray;

export const toCamel = (str: string): string => {
  return str.replace(/([-_][a-z])/gi, ($1) => {
    return $1.toUpperCase().replace("-", "").replace("_", "");
  });
};

export const isObject = (obj: Json): obj is JsonMap => {
  return obj === Object(obj) && !Array.isArray(obj);
};

export const snakeToCamel = (obj: Json): Json => {
  if (isObject(obj)) {
    const n: { [key: string]: Json } = {};

    Object.keys(obj).forEach((k) => {
      n[toCamel(k)] = snakeToCamel(obj[k]);
    });

    return n;
  }

  if (Array.isArray(obj)) {
    return obj.map((i) => {
      return snakeToCamel(i);
    });
  }

  return obj;
};
