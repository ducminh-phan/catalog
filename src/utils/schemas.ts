import { schema } from "normalizr";

export const category = new schema.Entity("categories");

export const categories = [category];

export const item = new schema.Entity("items");

export const items = [item];
