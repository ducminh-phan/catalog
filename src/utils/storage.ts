class Storage {
  private readonly prefix: string;

  private static readonly tokenKey: "token";

  constructor(localStoragePrefix: string) {
    this.prefix = localStoragePrefix;
  }

  getItem = (key: string): string | null =>
    localStorage.getItem(`${this.prefix}:${key}`);

  setItem = (key: string, value: string): void =>
    localStorage.setItem(`${this.prefix}:${key}`, value);

  removeItem = (key: string): void =>
    localStorage.removeItem(`${this.prefix}:${key}`);

  getToken = (): string | null => this.getItem(Storage.tokenKey);

  setToken = (token: string): void => this.setItem(Storage.tokenKey, token);

  removeToken = (): void => this.removeItem(Storage.tokenKey);
}

export default new Storage(process.env.REACT_APP_LOCAL_STORAGE_PREFIX);
