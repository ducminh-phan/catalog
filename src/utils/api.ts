import { Json, snakeToCamel } from "./misc";
import storage from "./storage";

type Method = "get" | "post" | "put";

class Request {
  static request = async <T extends Json>(
    endpoint: string,
    method: Method,
    body?: object,
  ): Promise<T> => {
    const token = storage.getToken();
    const headers: HeadersInit = {
      Accept: "application/json",
    };

    if (token) {
      headers.Authorization = `Bearer ${token}`;
    }

    const config: RequestInit = {
      method,
      headers,
    };

    if (body) {
      config.body = JSON.stringify(body);
      headers["Content-Type"] = "application/json";
    }

    const response = await window.fetch(
      `${process.env.REACT_APP_API_URL}/${endpoint}`,
      config,
    );

    const data = snakeToCamel(await response.json());

    if (response.ok) {
      return data as T;
    }

    return Promise.reject(data);
  };

  static get = <T extends Json>(endpoint: string): Promise<T> =>
    Request.request(endpoint, "get");

  static post = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "post", body);

  static put = <T extends Json>(endpoint: string, body: object): Promise<T> =>
    Request.request<T>(endpoint, "put", body);
}

export default Request;
