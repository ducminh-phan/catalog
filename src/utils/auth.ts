import Request from "./api";
import storage from "./storage";

export interface LoginRequest {
  username: string;
  password: string;
}

interface LoginResponse {
  access_token: string;
}

export interface RegisterRequest {
  username: string;
  password: string;
  email: string;
  name: string;
}

interface RegisterResponse {
  access_token: string;
}

const handleAuthResponse = ({
  access_token: token,
}: LoginResponse | RegisterResponse): void => storage.setToken(token);

export const login = (body: LoginRequest): Promise<void> => {
  return Request.post<LoginResponse>("login", body).then(handleAuthResponse);
};

export const register = (body: RegisterRequest): Promise<void> => {
  return Request.post<RegisterResponse>("registrations", body).then(
    handleAuthResponse,
  );
};

export const logout = (): Promise<void> => {
  storage.removeToken();
  return Promise.resolve();
};

interface User {
  id: number;
  name: string;
}

export interface AppData {
  user: User;
}

interface UserResponse {
  email: string;
  username: string;
  created: string;
  updated: string;
  name: string;
  id: number;
}

export const getAppData = async (): Promise<AppData | null> => {
  const token = storage.getToken();

  if (token === null) {
    return null;
  }

  return Request.get<UserResponse>("me")
    .then(({ id, name }) => ({ user: { id, name } }))
    .catch((error) => {
      logout();
      return Promise.reject(error);
    });
};
