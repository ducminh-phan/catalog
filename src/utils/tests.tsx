import { StylesProvider } from "@material-ui/styles";
import { render as rtlRender, RenderResult } from "@testing-library/react";
import { Rule, StyleSheet } from "jss";
import React, { ComponentType, ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";

import Notification from "components/Notification";
import { AuthProvider } from "contexts/auth";
import { NotificationProvider } from "contexts/notification";
import store from "store";

import Request from "./api";
import { UserResponse } from "./auth";
import storage from "./storage";

// Custom logic to make Material UI class names consistent in snapshots
const generateClassName = (
  rule: Rule,
  styleSheet?: StyleSheet<string>,
): string => `${styleSheet?.options.classNamePrefix}-${rule.key}`;

const render = (ui: ReactElement, renderOptions = {}): RenderResult => {
  const Wrapper = ({ children }: { children: ReactNode }): ReactElement => {
    return (
      <StylesProvider generateClassName={generateClassName}>
        <Provider store={store}>
          <NotificationProvider>
            <AuthProvider>{children}</AuthProvider>
            <Notification />
          </NotificationProvider>
        </Provider>
      </StylesProvider>
    );
  };

  return rtlRender(ui, { wrapper: Wrapper as ComponentType, ...renderOptions });
};

export const setupUser = (): void => {
  storage.setToken("accessToken");
  expect(storage.getToken()).toBe("accessToken");
  jest.spyOn(Request, "get").mockImplementation(
    (): Promise<UserResponse> =>
      Promise.resolve({
        email: "minhπ",
        username: "minhπ",
        created: "",
        updated: "",
        name: "minhπ",
        id: 0,
      }),
  );
};

// re-export everything
export * from "@testing-library/react";

// override render method
export { render };
