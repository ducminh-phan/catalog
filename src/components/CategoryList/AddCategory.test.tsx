import React from "react";

import store from "store";
import Request from "utils/api";
import {
  fireEvent,
  render,
  setupUser,
  wait,
  waitForElementToBeRemoved,
} from "utils/tests";
import * as types from "utils/types";

import AddCategory from "./AddCategory";

it("can add new category", async () => {
  setupUser();
  const {
    baseElement,
    getByLabelText,
    queryByLabelText,
    getByRole,
    getByText,
    queryByText,
    getByTestId,
  } = render(<AddCategory />);

  await wait(() => getByTestId("add-category"));
  expect(queryByText("Cancel")).toBeNull();

  const addButton = getByTestId("add-category");

  // Click the Add button
  fireEvent.click(addButton);
  await wait(() => getByText("Cancel"));
  expect(queryByLabelText(/Name.*/)).toBeInTheDocument();
  expect(queryByLabelText(/Description.*/)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();

  // Click the Cancel button
  fireEvent.click(getByText("Cancel"));
  await waitForElementToBeRemoved(() => queryByText("Cancel"));
  expect(queryByLabelText(/Name.*/)).toBeNull();
  expect(queryByLabelText(/Description.*/)).toBeNull();

  // Click the Add button again
  fireEvent.click(addButton);
  await wait(() => getByText("Cancel"));

  // The submit button should be disabled
  const submitButton = [
    ...getByRole("presentation").querySelectorAll("button"),
  ].filter((node) => node.textContent === "Submit")[0];
  expect(submitButton).toHaveAttribute("disabled");

  // Fill in mock values
  const nameField = getByLabelText(/Name.*/) as HTMLInputElement;
  const descriptionField = getByLabelText(/Description.*/) as HTMLInputElement;
  const mockId = 0;
  const mockName = "mockName";
  const mockDescription = "mockDescription";
  const mockCategory: types.Category = {
    id: mockId,
    name: mockName,
    description: mockDescription,
  };
  await wait(() => {
    fireEvent.change(nameField, {
      target: { value: mockName },
    });
    fireEvent.change(descriptionField, {
      target: { value: mockDescription },
    });
  });

  expect(nameField.value).toBe(mockName);
  expect(descriptionField.value).toBe(mockDescription);
  expect(submitButton).not.toHaveAttribute("disabled");

  // Submit
  jest
    .spyOn(Request, "post")
    .mockImplementation(
      (): Promise<types.Category> => Promise.resolve(mockCategory),
    );

  await wait(() => fireEvent.click(submitButton));
  expect(Request.post).toHaveBeenCalledWith("categories", {
    name: mockName,
    description: mockDescription,
  });

  const { categories } = store.getState().category;
  expect(Object.keys(categories)).toHaveLength(1);
  expect(categories[mockId.toString()]).toEqual(mockCategory);
});
