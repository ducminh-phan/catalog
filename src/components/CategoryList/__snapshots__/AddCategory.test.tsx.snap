// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new category 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-colorInherit"
      data-testid="add-category"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
  <div
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="category-add-form-title"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
          id="category-add-form-title"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            Add Category
          </h2>
        </div>
        <div
          class="MuiDialogContent-root"
        >
          <form
            novalidate=""
          >
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiFormLabel-required MuiInputLabel-required"
                data-shrink="false"
                for="add-category-name"
                id="add-category-name-label"
              >
                Name
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="add-category-name"
                  name="name"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiFormLabel-required MuiInputLabel-required"
                data-shrink="false"
                for="add-category-description"
                id="add-category-description-label"
              >
                Description
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   
                  *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  id="add-category-description"
                  name="description"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="MuiDialogActions-root MuiDialogActions-spacing"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Cancel
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-disabled MuiButtonBase-disabled"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
